### ISSUE TITLE
distutils is removed in Python 3.12

### ISSUE THREAD

**Issue by cclauss** [nodejs/node-gyp#2869]
[distutils](https://docs.python.org/3/library/distutils.html#module-distutils) is deprecated with removal planned for Python 3.12. See the [What’s New](https://docs.python.org/3/whatsnew/3.10.html#distutils-deprecated) entry for more information.

Our only current use of [Python `distutils`](https://docs.python.org/3/library/distutils.html) is at:
https://github.com/nodejs/node-gyp/blob/33391db3a0008eff8408890da6ab232f2f90fcab/gyp/pylib/gyp/input.py#L19
https://github.com/nodejs/node-gyp/blob/33391db3a0008eff8408890da6ab232f2f90fcab/gyp/pylib/gyp/input.py#L1186

As discussed in https://stackoverflow.com/questions/11887762, we should replace `distutils.version.StrictVersion` with [`packaging.version.parse`](https://packaging.pypa.io/en/latest/version.html#packaging.version.parse).

**Comment by cclauss** [comment]
@rzhao271 or @dhruvmanila -- Any interest in fixing this one?

**Comment by rzhao271** [comment]
I'm interested, though it'll have to be on my own time.

**Comment by cclauss** [comment]
#2888 Needs help to pass test…

**Comment by cclauss** [comment]
This is becoming more urgent now that **Python 3.12 is a release candidate**.
* https://www.python.org/download/pre-releases
*  #2862 demonstrates the problem
* #2888 is not yet a solution

**Comment by sibelius** [comment]
this is breaking for me

**Comment by cclauss** [comment]
Py3.12 is still pre-release.  Please try to copy the `pylib/gyp/input.py` changes from nodejs#201 and let us know if that unblocks you.

**Comment by rzhao271** [comment]
The upstream fix still hasn't been merged https://github.com/nodejs/gyp-next/pull/197

**Comment by Steveorevo** [comment]
This is also broken for me. Simply create an electron forge app and then attempt to use like:

```
require('deasync').sleep(1000);
console.log('Hello world');
```

I'm greeted with:

```
File node_modules\node-gyp\gyp\pylib\gyp\input.py", line 19, in <module>
  from distutils.version import StrictVersion
  ModuleNotFoundError: No module named 'distutils'
  Error: `gyp` failed with exit code: 1
  
An unhandled rejection has occurred inside Forge:
Error: node-gyp failed to rebuild '... \resources\app\node_modules\deasync'
```

**Comment by cclauss** [comment]
@rzhao271 @dhruvmanila This is still broken on Windows see these attempts to fix https://github.com/nodejs/node-gyp/issues/2869#issuecomment-1693409123

**Comment by jurijsk** [comment]
Python 3.12 released on Oct 2nd and got it by default with other tooling and nothing is working now because `distutils` is removed

**Comment by nikolay** [comment]
@jurijsk It's sad, as we knew Python was removing it and have done nothing about it!

**Comment by pplmx** [comment]
Any progress? node-gyp cannot work on Python 3.12 now.

**Comment by GorlikItsMe** [comment]
Same issue here (Windows, python 3.12.0)

**Comment by birdzai** [comment]
Doesn't work on Mac either

**Comment by iki** [comment]
> The upstream fix still hasn't been merged [nodejs/gyp-next#197](https://github.com/nodejs/gyp-next/pull/197)

@rzhao271 upstream was [merged](https://github.com/nodejs/gyp-next/pull/197#issuecomment-1744110600) since that

> 🤖 Release is at https://github.com/nodejs/gyp-next/releases/tag/v0.15.1 🌻

**Comment by cclauss** [comment]
I will work on this today.

**Comment by cclauss** [comment]
If some of you are frustrated then please do `python3 -m pip install packaging` and report back here if that fixes things.

Other activities blocked vendoring effort today but I will dig in with my first coffee tomorrow.

**Comment by nzikic** [comment]
@cclauss This fixed the issue! Thank you.

**Comment by cclauss** [comment]
Thx.  I will make it so that step is not needed but your effort really helps.

**Comment by confused-Techie** [comment]
I'm seeing errors in our build process that `distutils` is missing in Python `3.11.6`, I can't find reference to these changes being backported in the changelog for this Python version, but maybe it's possible?

<details>
<summary>Partial Log</summary>

```
D:\a\pulsar\pulsar\ppm\node_modules\git-utils>if not defined npm_config_node_gyp (node "D:\a\pulsar\pulsar\ppm\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "D:\a\pulsar\pulsar\ppm\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js" rebuild ) 
undefined
Traceback (most recent call last):
  File "D:\a\pulsar\pulsar\ppm\node_modules\npm\node_modules\node-gyp\gyp\gyp_main.py", line 42, in <module>
    import gyp  # noqa: E402
    ^^^^^^^^^^
  File "D:\a\pulsar\pulsar\ppm\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\__init__.py", line 9, in <module>
    import gyp.input
  File "D:\a\pulsar\pulsar\ppm\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\input.py", line 19, in <module>
    from distutils.version import StrictVersion
ModuleNotFoundError: No module named 'distutils'
gyp ERR! configure error 
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack at ChildProcess.onCpExit (D:\a\pulsar\pulsar\ppm\node_modules\npm\node_modules\node-gyp\lib\configure.js:305:16)
gyp ERR! stack at ChildProcess.emit (node:events:365:28)
gyp ERR! stack at Process.ChildProcess._handle.onexit (node:internal/child_process:290:12)
gyp ERR! System Windows_NT 10.0.20348
gyp ERR! command "D:\\a\\pulsar\\pulsar\\ppm\\bin\\node.exe" "D:\\a\\pulsar\\pulsar\\ppm\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd D:\a\pulsar\pulsar\ppm\node_modules\git-utils
gyp ERR! node -v v16.0.0
gyp ERR! node-gyp -v v9.4.0
gyp ERR! not ok 
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! git-utils@5.7.3 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the git-utils@5.7.3 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\npm\cache\_logs\2023-10-18T18_54_57_896Z-debug.log
Done in 208.45s.
Command completed after 1 attempt(s).
```

</details>

Or here's the full run where this occurs:
* [Using Python 3.11.6](https://github.com/pulsar-edit/pulsar/actions/runs/6570347369/job/17847608573) - Where it fails
* [Using Python 3.11.5](https://github.com/pulsar-edit/pulsar/actions/runs/6542424454/job/17765426913) - Where it succeeds

**Comment by cclauss** [comment]
The Py3.11 runs that you linked to all pass, not fail.

`distutils` was removed in Py3.12 but 3.11 should be fine.  If you can reproduce on Py3.11, please open a SEPARATE issue and include a FULL log.

The "Partial" Log is a mistake because the lines just above the "Partial" part would have documented exactly which Python version was running.  My bet is that it was Py3.12

**Comment by confused-Techie** [comment]
@cclauss I understand that it reports it passed, but the logs do include a failure. Apologies. I'll open a separate issue.

**Comment by kjackson1998** [comment]
> I'm seeing errors in our build process that `distutils` is missing in Python `3.11.6`, I can't find reference to these changes being backported in the changelog for this Python version, but maybe it's possible?
> 
> Partial Log
> Or here's the full run where this occurs:
> 
> * [Using Python 3.11.6](https://github.com/pulsar-edit/pulsar/actions/runs/6570347369/job/17847608573) - Where it fails
> * [Using Python 3.11.5](https://github.com/pulsar-edit/pulsar/actions/runs/6542424454/job/17765426913) - Where it succeeds
> 
> EDIT: I may have spoken too soon. Pinning Python to `3.11.5` still results in `No module named 'distutils'`. Here's the [run](https://github.com/pulsar-edit/pulsar/actions/runs/6581435641/job/17881309220?pr=779). So not sure if this is the same issue or not, apologies if not and this is just noise. Just seems interesting the error matches exactly.
> 
> But I'll try the command suggested above to install `packaging` or the Python [Changelog](https://docs.python.org/dev/whatsnew/3.12.html#distutils) suggests installing `setuptools` to reintroduce `distutils`.

for me, fwiw:

pip install packaging - did not work
pip install setuptools - worked

**Comment by cclauss** [comment]
# Update: Today `gyp-next v0.16.0` was released.
* https://github.com/nodejs/gyp-next/releases

This includes some key Windows and Python 3.12 fixes.
Once that is brought into this repo we should see nice improvements in all the issues discussed above.

Who has the expertise to resync with the current `gyp-next`.

**Comment by martin-braun** [comment]
> > EDIT: This advice was for `node-gyp` < v10 but is not required for later versions.
> 
> If some of you are frustrated then please do `python3 -m pip install packaging` and report back here if that fixes things.
> 
> Other activities blocked vendoring effort today but I will dig in with my first coffee tomorrow.

Wasn't working for me in the case of installing `@signalapp/better-sqlite3`, but doing `pip install setuptools` solved it for me.

https://github.com/WiseLibs/better-sqlite3/issues/1093

**Comment by voidmain443** [comment]
I am using python3.12 for venv and I have a same problem with execute pandas_datareader.

**Comment by cclauss** [comment]
Please look at your error log and ensure it contains the line [`gyp info using node-gyp@10.0.1`](https://github.com/nodejs/node-gyp/releases).

**Comment by federicoWax** [comment]
> pip install setuptools

Tengo dos semanas buscando el bug porque no podia correr tensorflow/tfjs muchas gracias!

**Comment by RubTalha** [comment]
https://stackoverflow.com/a/76691103/13086128

**Comment by cclauss** [comment]
EDIT: This issue was **_closed_** via nodejs/gyp-next#214 which is available in [node-gyp v10+](https://github.com/nodejs/node-gyp/releases).
* nodejs/gyp-next#214 -->
* #2923
